cloudinit/net/netplan.py

```diff
@@ -287,22 +344,24 @@ def render_network_state(
             header += "\n"
         content = header + content
 
-        # determine if existing config files have the same content
         same_content = False
-        if os.path.exists(fpnplan):
-            hashed_content = util.hash_buffer(io.BytesIO(content.encode()))
-            with open(fpnplan, "rb") as f:
-                hashed_original_content = util.hash_buffer(f)
-            if hashed_content == hashed_original_content:
-                same_content = True
-
-        mode = 0o600 if features.NETPLAN_CONFIG_ROOT_READ_ONLY else 0o644
-        if not same_content and os.path.exists(fpnplan):
-            current_mode = util.get_permissions(fpnplan)
-            if current_mode & mode == current_mode:
-                # preserve mode if existing perms are more strict than default
-                mode = current_mode
-        util.write_file(fpnplan, content, mode=mode)
+        if not netplan_api_write_yaml_file(content):
+            # No netplan api, fallback to write config  directly.
```

Comment:
Nit: As we're adding some extra complexity here, I think the code under this `if block` would be better as its own function. `fallback_write_yaml_file` or something like that.